File: online-judges/leetcode/is-subsequence.rs
// https://leetcode.com/problems/is-subsequence/

impl Solution {
    pub fn is_subsequence(s: String, t: String) -> bool {
        let mut cur = 0;

        for c in t.chars() {
            if cur >= s.len() {
                return true;
            }
            if s.chars().nth(cur) == Some(c) {
                cur += 1;
            }
        }

        cur >= s.len()
    }
}
//

struct Solution {}

#[test]
fn test1() {
    let s = "abc".to_string();
    let t = "ahbgdc".to_string();
    let is_subsequence = Solution::is_subsequence(s, t);
    assert!(is_subsequence);
}

#[test]
fn test2() {
    let s = "axc".to_string();
    let t = "ahbgdc".to_string();
    let is_subsequence = Solution::is_subsequence(s, t);
    assert!(!is_subsequence);
}

#[test]
fn test3() {
    let s = "".to_string();
    let t = "ahbgdc".to_string();
    let is_subsequence = Solution::is_subsequence(s, t);
    assert!(is_subsequence);
}

#[test]
fn test4() {
    let s = "b".to_string();
    let t = "c".to_string();
    let is_subsequence = Solution::is_subsequence(s, t);
    assert!(!is_subsequence);
}

File: online-judges/leetcode/is-subsequence.py
# https://leetcode.com/problems/is-subsequence/


class Solution:
    def isSubsequence(self, s: str, t: str) -> bool:
        i = 0
        j = 0
        while i < len(s) and j < len(t):
            if s[i] == t[j]:
                i += 1
            j += 1
        return i == len(s)


# Tests
solver = Solution()
assert solver.isSubsequence("abc", "ahbgdc") == True
assert solver.isSubsequence("axc", "ahbgdc") == False
